### PR TITLE
Adding ArmException implementation

### PR DESCRIPTION
**Pull request by allenjzhang**

### CODE REVIEW

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    /// <summary>
+    /// A class representing exceptions thrown by Azure ARM services.
+    /// </summary>
+    [Serializable]
+    public class ArmException : RequestFailedException
+    {
+        /// <summary>
+        /// Gets and set the Code of the ARM exception.
+        /// </summary>
+        public string Code { get; internal set; }
```
Comment (pakrym) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs:22]: RequestFailedException already has an `ErrorCode` property having both might be confusing.

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+        /// <summary>
+        /// Gets and set the Details of the ARM exception.
+        /// </summary>
+        public ArmException[]? Details { get; internal set; }
+
+        /// <summary>
+        /// Gets and set the AdditionalInfo of the ARM exception.
+        /// </summary>
+        public IDictionary AdditionalInfo => Data;
```
Comment (pakrym) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs:37]: I don't think we generally add aliases for existing properties especially if they don't do any transformations.

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+        /// <summary>
+        /// Gets and set the Target of the ARM exception.
+        /// </summary>
+        public string Target { get; internal set; }
+
+        /// <summary>
+        /// Gets and set the Details of the ARM exception.
+        /// </summary>
+        public ArmException[]? Details { get; internal set; }
```
Comment (pakrym) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs:32]: Can we represent this using the InnerException + AggregateException type?

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/Resources/ErrorAdditionalInfo.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+// Licensed under the MIT License.
+
+// <auto-generated/>
+
+#nullable disable
+
+namespace Azure.ResourceManager.Core.Resources
+{
+    /// <summary> The resource management error additional info. </summary>
+    public partial class ErrorAdditionalInfo
```
Comment (pakrym) [inline sdk/resourcemanager/Azure.ResourceManager.Core/src/Resources/ErrorAdditionalInfo.cs:11]: Do we want to expose this type along with ErrorResponse?

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/ArmExceptionTests.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+using Azure.Core.Pipeline;
+using Moq;
+using Moq.Protected;
+using NUnit.Framework;
+
+#nullable enable
+
+namespace Azure.ResourceManager.Core.Tests.Unit
+{
+    public class ArmExceptionTests
```
Comment (pakrym) [inline sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/ArmExceptionTests.cs:20]: 👍

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/ArmExceptionTests.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        private ClientDiagnostics GetClientDiagnostics()
+        {
+            var options = new Mock<ClientOptions>();
+
+            return new ClientDiagnostics(options.Object);
+        }
+
+        private Response GetMockResponse(string jsonAssetFileName)
+        {
+            var mock = new Mock<Response>();
```
Comment (pakrym) [inline sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/ArmExceptionTests.cs:76]: Can you use the `MockResponse` type here?

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs (hunk 1 above)
Comment (allenjzhang) [reply sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs:22]: Arm Error (Odata error) has Code, Message etc. It is direct mapping. I would have preferred the RequestFailed to be HttpStatusCode, but...

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs (hunk 3 above)
Comment (allenjzhang) [reply sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs:32]: Sounds good.

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs (hunk 2 above)
Comment (allenjzhang) [reply sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs:37]: I was debating that one. Was hoping to match exactly on the ErrorResponse so user can have clear linkage between the C# exception and Rest contract. Same for the Details. Should discuss on this.

File: sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs (hunk 3 above)
Comment (allenjzhang) [reply sdk/resourcemanager/Azure.ResourceManager.Core/src/ArmException.cs:32]: Implementation do-able. Though need decision on the point raised on AdditionalInfo as well.

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/ArmExceptionTests.cs (hunk 6 above)
Comment (allenjzhang) [reply sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/ArmExceptionTests.cs:76]: Doh, could have saved me some time to get the Mock.Setup all working. :-)

File: sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/TestAssets/ArmException/ArmExceptionSimpleV1.json [hunk 7]
```diff
@@ -0,0 +1,24 @@
+{
+    "code": "400",
```
Comment (m-nash) [inline sdk/resourcemanager/Azure.ResourceManager.Core/tests/Unit/TestAssets/ArmException/ArmExceptionSimpleV1.json:2]: should we have a few more cases like

- required things missing and assert the proper serialization failure
- missing opitional elements and assert nulls / empties
- etc

**Review by m-nash — APPROVED** [review]

**Comment by allenjzhang** [comment]
Closing this PR for now as similar feature is planned on RequestFailedException

**Pull request closed without merging**